R/tables.R

```diff
@@ -1,117 +1,274 @@
+#'
+#' Note that subsetting in SQL is more limited:
+#'
+#' -  Use `0` or `1` instead of `TRUE` or `FALSE`
+#' -  Dates are stored as character strings, but as they are in YYYY-MM-DD
+#'    format, filtering such as `event_date > "2014-01-01"` still works.
+#'
+#' @param table one of the EIDITH database tables. One of "events", "animals",
+#' "specimens", "tests", "viruses", or "test_specimen_ids".
+#' @param ... arguments passed to [dplyr::filter()] to subset data
+#' @param .dots standard-evaluation versions of subsetting arguments
+#' @return a [tibble][tibble::tibble]-style data frame.
+#' @importFrom dbplyr partial_eval
+#' @importFrom dplyr tbl tbl_df filter_ mutate_at funs_ funs collect vars
+#' @importFrom stringi stri_replace_first_regex stri_extract_last_regex stri_detect_fixed
+#' @export
+#' @rdname ed_table
+ed_table_ <- function(table, ..., .dots) {
+  dots <- lazyeval::all_dots(.dots, ...)
+  dots = lazyeval::as.lazy_dots(   #This stuff deals with the dplyr bug found at https://github.com/hadley/dplyr/issues/511 by modifying "%in%" calls
+    lapply(dots, function(dot_expr) {
+      new_expr <- paste0(
+        deparse(partial_eval(dot_expr[["expr"]], env=dot_expr[["env"]]),
+                width.cutoff = 500L),
+        collapse = "")
+      if(stri_detect_fixed(new_expr, "%in%")) {
+        matched_expr <- stri_extract_last_regex(new_expr, "(?<=%in%\\s).*$")
+        if(length(eval(parse(text=matched_expr))) ==  0 ) {
+          new_expr <- stri_replace_first_fixed(new_expr, matched_expr, "('')")
+        }
+        else if(length(eval(parse(text=matched_expr))) == 1) {
+          new_expr <- stri_replace_first_fixed(new_expr, matched_expr,
+                                               paste0("(", matched_expr, ")"))
+        }
+      }
+      lazyeval::as.lazy(new_expr, env=dot_expr[["env"]])
+    }))
+
+
+  #add class attributes to P2 tables
+  if(stri_detect_fixed(table, "2")){
+    ed_tb <- dbReadTable(eidith_db()$con, table)
+    #adding notes
+    note_cols <- which(stri_detect_fixed(names(ed_tb), "notes"))
+    attr(ed_tb, "notes") <- note_cols
+    #adding duplicate rows
+    intended_key <- db_other_indexes[[table]][[1]]
+    group_var <- as.name(intended_key)
+    key_errors <- ed_tb %>%
+      group_by(!!group_var) %>%
+      summarize(count = n()) %>%
+      filter(count > 1) %>%
+      dplyr::pull(!!group_var)
+    attr(ed_tb, "duplicate_keys") <- key_errors
+    indices <- which(ed_tb[[intended_key]] == key_errors)
+    attr(ed_tb, "duplicate_indices") <- indices
+    class(ed_tb) <- c("eidith_tbl", class(ed_tb))
+
+    #print details on local database pull"
+    if(interactive()){
+    if(length(note_cols > 0)){
+      cat_line("")
+      cat_line(cyan("There are notes attached to this dataframe which may contain important information! Please look at the following column(s):"))
+      cat_line(paste("    ", magenta(names(ed_tb)[note_cols])))
+    }

```

Comment:
Can we report the number of rows with notes in them?

Reply:
Done.